### ISSUE TITLE
New Rule Proposal: react/prefer-pure-calc-over-effect

### ISSUE THREAD

**Issue by stevenhankin** [yannickcr/eslint-plugin-react#2936]
This isn't my idea (see links in footer) but I couldn't see this rule in the backlog and looks as though it could be useful.
The rule would indicate where a useEffect() has been used to make a calculation and call a setter, but which could and should be replaced with a useMemo()

Example of **incorrect** code:

```js
const [processedData, setProcessedData] = useState();

useEffect(() => {
  let processed = /* do something with data */;
  setProcessedData(processed);
}, [data]);
```

Example of **correct** code:

```js
const processedData = useMemo(() => {
  let processed = /* do something with data */;
  return processed;
}, [data]);
```

See [Sophie's Tweet](https://twitter.com/sophiebits/status/1293710971274289152?ref_src=twsrc%5Etfw) and [here](https://javascript.plainenglish.io/react-usememo-and-when-you-should-use-it-e69a106bbb02)

**Comment by ljharb** [comment]
There's a critical difference there tho - useEffect isn't run on the server, while useMemo is. I don't think it's ever going to be generically safe to suggest changing a browser-only hook into a universal one.

**Comment by stevenhankin** [comment]
I think I see what you mean, although if the state that is being derived is based on (and _only_ on) an input that has triggered the hook, then it should really be a Memo to make it a pure function?  Then regardless of it executing on Server or Client, the result is the same?

**Comment by ljharb** [comment]
That depends on if the input is something that can happen on the server or not.

**Comment by stevenhankin** [comment]
Yes, that could be a problem. I'll close this proposal 😺
